### ISSUE TITLE
[0.12.0] Goroutine leak with intense write load and shard in state "cache maximum memory size exceeded"

### ISSUE THREAD

**Issue by mark-rushakoff** [influxdata/influxdb#6417]
### Bug report

__System info:__ 

InfluxDB 0.12.0, Linux

__Steps to reproduce:__

Apply an intense write load.

__Expected behavior:__ 

Shard writes succeed.

__Actual behavior:__ 

Log lines:

The line occurs once early in the logs:

```
[store] 2016/04/19 01:22:31 Failed to open shard: 1: [shard 1] cache maximum memory size exceeded
```

Then repeatedly throughout the rest of the logs, this line appears:

```
[cluster] 2016/04/19 01:22:41 process write shard error: create shard 1: [shard 1] cache maximum memory size exceeded
```

Eventually the system OOMs, and the following goroutine stacks are repeated many times:

```
goroutine 48 [semacquire, 9 minutes]:
sync.(*Mutex).Lock(0xc208038380)
        /usr/local/go/src/sync/mutex.go:66 +0xd3
sync.(*RWMutex).Lock(0xc208038380)
        /usr/local/go/src/sync/rwmutex.go:82 +0x33
github.com/influxdata/influxdb/tsdb.(*Store).CreateShard(0xc208038380, 0xc22d193400, 0xe, 0xc22d1933b8, 0x7, 0x1, 0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/tsdb/store.go:252 +0x59
github.com/influxdata/influxdb/cluster.(*Service).processWriteShardRequest(0xc2080a40e0, 0xc231c0c000, 0xe693, 0xe693, 0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/service.go:273 +0x336
github.com/influxdata/influxdb/cluster.(*Service).handleConn(0xc2080a40e0, 0x7f2d0d957388, 0xc208036050)
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/service.go:182 +0x780
github.com/influxdata/influxdb/cluster.func·005()
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/service.go:126 +0x74
created by github.com/influxdata/influxdb/cluster.(*Service).serve
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/service.go:127 +0x428
```

```
goroutine 7186 [select]:
github.com/influxdata/influxdb/cluster.func·006()
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/service.go:150 +0xff
created by github.com/influxdata/influxdb/cluster.(*Service).handleConn
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/service.go:155 +0x188
```

__Additional info:__

Full logs here: [log.gz](https://github.com/influxdata/influxdb/files/225127/log.gz)

**Comment by mark-rushakoff** [comment]
See also https://github.com/influxdata/influxdb/issues/6109.

**Comment by jwilder** [comment]
@mark-rushakoff Can you try with `0.12.1` or `master`?  #6246 was backported to 0.12 after `eb467d8d` which changed some of the locking related to this code.  The logs you uploaded don't look like they have those changes.

**Comment by jwilder** [comment]
It looks like when the server is started and it tries to reload shard 1's data into the cache, it's hitting the `cache maximum memory size exceeded` limit for some reason and aborting loading that shard.  The server then comes up and remote writes start coming in for that shard and each one tries to reload the shard into the cache before accepting the write.  They fail again with the cache memory error, but the writes come in faster than the shard loading process and pile up.  Those goroutines are getting blocked trying to acquire `tsdb.Store` write lock and each one of the connections may have a large point payload that eventually causes the OOM.  

I'm not sure that the locking changes in #6246 will prevent this.

A workaround would be to bump up `cache-max-memory-size` in the config to allow the shard to load at startup correctly.  Not sure why it was able to write the points initially the first time, but fails to reload them.

We may need to disable the `cache-max-memory-size` setting when reloading a shard or record that the shard failed to open and don't keep trying to re-open in when writes come in so they fail fast with an error.

**Comment by kirillkovalenko** [comment]
We've just upgraded from 0.10.x to v0.12.2 and observe massive [go-routine leakage](https://onedrive.live.com/redir?resid=9844FDC23B276A96!4321&authkey=!AAnsrLXVzpr7Sn4&v=3&ithint=photo%2cpng) as well. I would not call our case write intensive though, we have ~1gb of data per week.

**Comment by jwilder** [comment]
@kirillkovalenko How are you writing data?  Using telegraf, something else?

Would you be able to get a stack trace?  If you send a `SIGQUIT` signal to the process when you observe the large number of goroutines, it will dump a stack trace to the logs.  That would be very useful.

**Comment by mark-rushakoff** [comment]
@kirillkovalenko are you seeing a line like this in your logs?

```
[cluster] 2016/04/19 01:22:41 process write shard error: create shard 1: [shard 1] cache maximum memory size exceeded
```

If so, that's consistent with the original bug in this issue – something is causing a shard to exceed its memory limit, which seems to result in a lock not being release, which seems to cause goroutines to queue up indefinitely as writes continue.

If you don't see a line like the one above, it's probably a separate issue. Either way, +1 to @jwilder's suggestion of SIGQUIT so we can see exactly what the goroutines are doing.

**Comment by jwilder** [comment]
@mark-rushakoff Are you still able to reproduce the write shard error w/ current master?

**Comment by mark-rushakoff** [comment]
@jwilder I'll try to get an answer later today.

**Comment by kirillkovalenko** [comment]
@jwilder we are using influx-capacitor and a custom app based on the library that influx-capacitor uses.

**Comment by kirillkovalenko** [comment]
@jwilder [Here are the logs](https://onedrive.live.com/redir?resid=9844FDC23B276A96!4337&authkey=!AP4KpUFCE5AAQ20&ithint=file%2czip) you've been asking for.

**Comment by kirillkovalenko** [comment]
@mark-rushakoff  In our logs I was not able to find this error.

**Comment by kirillkovalenko** [comment]
@jwilder  JFYI, after restarting the whole thing (influxdb, grafana, agents) we [see the leak no more](https://onedrive.live.com/redir?resid=9844FDC23B276A96!4343&authkey=!APyUEMLR5gbTR2k&v=3&ithint=photo%2cpng).

**Comment by jwilder** [comment]
@kirillkovalenko Your logs are different than this issue.  You have a lot of open HTTP connections which looks similar to #6449.  I'm not familiar with influx-capacitor, but it looks like one of your tools may not be closing connections on the client side.  Telegraf and possibly [intel snap influxdb plugin](https://github.com/influxdata/influxdb/issues/6185#issuecomment-215503044) had similar issues.  They were both using the go client which was fixed with #6425.

**Comment by kirillkovalenko** [comment]
@jwilder thanks for your help. i'll continue investigating on the client side.
